modules/java/generator/src/cpp/Mat.cpp

```diff
@@ -2159,16 +2159,27 @@ template<typename T> static int mat_put(cv::Mat* m, int row, int col, int count,
     return res;
 }
 
-// returns true if final index was reached
-static bool updateIdx(cv::Mat* m, std::vector<int>& idx, int inc) {
-    for (int i=m->dims-1; i>=0; i--) {
-        if (inc == 0) return false;
-        idx[i] = (idx[i] + 1) % m->size[i];
-        inc--;
+// return true if we have reached the final index
+static bool incIdx(cv::Mat* mat, std::vector<int>& indices) {
+    for (int dim = mat->dims-1; dim>=0; dim--) {
+        indices[dim] = (indices[dim] + 1) % mat->size[dim];
+        if (indices[dim] != 0) {
+            return false;
+        }
     }
     return true;
 }
 
+// returns true if final index was reached
+static bool updateIdx(cv::Mat* mat, std::vector<int>& indices, int inc) {
+    for (int index = 0; index < inc; index++) {
+        if (incIdx(mat, indices)) {
+            return true;
```

Comment:
Increasing by one multiple times is really slow (for HD images `inc` value is up to 1280).
We need "fast path" to handle "always contiguous" last dim (cols).

```
{
    int cols_dim = mat->dims - 1;
    size_t cols = mat->size[cols_dim];
    CV_CheckLE(cols, (size_t)INT_MAX, "");  // 'int' only, size_t type is not supported for indexes
    size_t new_cols_idx = (size_t)indices[cols_dim] + inc;
    // only "cols" idx is allowed to be increased till the end
    // do not allow to jump forward because we should support non-contiguous layouts too (submats)
    CV_CheckLE(new_cols_idx, cols, "");
    if (new_cols_idx == cols)
    {
        indices[cols_dim] = 0;
        for (int dim = cols_dim - 1; dim >= 0; --dim)
        {
            int new_idx = indices[dim] + 1;
            if (new_idx != mat->size[dim])
            {
                indices[dim] = new_idx;
                return false;
            }
            else
            {
                indices[dim] = 0;
                continue;
            }
        }
        return true;
    }
    else
    {
        indices[cols_dim] = (int)new_cols_idx;
        return false;
    }
}

```

Reply:
Actually updating the indices doesn't require any iteration - we can just calculate where the new indices will end up.
But I get your point - we can optimize the copying of data if the trailing dimensions are contiguous. Will push a new optimized implementation.